Unify Stark and Secp256 Curves interfaces (#1352)
foundry-rs/starknet-foundry: docs/src/appendix/forge-library/signature/secp256_curve.md

# `Secp256 Curve`

This module contains the implementation of `KeyPairTrait` for the `Secp256k1` and `Secp256r1` curves.

> > ⚠️ **Security Warning**
> >
> > Please note that some parts of the internal code for the features in this module are still experimental and **have not been audited yet**.
> >
> > Use at your own risk!

Reviewer: ❤️ 
```suggestion
> ⚠️ **Security Warning**
>
> Please note that cryptography in Starknet Foundry is still experimental and **have not been audited yet**.
>
> Use at your own risk!
```

## `Secp256CurveKeyPairImpl`

The `Secp256CurveKeyPairImpl` is an implementation of `KeyPair` with secret key being `u256` and private key being a `Secp256Point`.

The `Secp256Point` implements `Secp256PointTrait` from corelib.

```rust
struct KeyPair {
    secret_key: u256,
    public_key: Secp256Point,
}
```

A new key pair for the `Secp256r1` curve can be obtained as follows:

```rust
use snforge_std::signature::{KeyPair, KeyPairTrait, SignerTrait, VerifierTrait};
use snforge_std::signature::secp256_curve::{Secp256CurveKeyPairImpl, Secp256CurveSignerImpl, Secp256CurveVerifierImpl};
use starknet::secp256r1::{Secp256r1Impl, Secp256r1Point, Secp256r1PointImpl};

let key_pair_secp256r1 = KeyPairTrait::<u256, Secp256r1Point>::generate();
```

Or, if the `secret_key` is known in advance:

```rust
use snforge_std::signature::{KeyPair, KeyPairTrait, SignerTrait, VerifierTrait};
use snforge_std::signature::secp256_curve::{Secp256CurveKeyPairImpl, Secp256CurveSignerImpl, Secp256CurveVerifierImpl};
use starknet::secp256r1::{Secp256r1Impl, Secp256r1Point, Secp256r1PointImpl};

let secret_key = ...;
let key_pair_secp256r1 = KeyPairTrait::<u256, Secp256r1Point>::from_secret_key(secret_key);

> ```

Reviewer: Please drop this part, it is unnecessary wordy, global example + signatures are enough.

## `Secp256CurveSignerImpl`

Implementation of the `SignerTrait` for the Secp256 curve.

> fn sign(self: KeyPair<u256, Secp256Point>, message_hash: H) -> (u256, u256)

## `Secp256CurveVerifierImpl`

Implementation of the `VerifierTrait` for the Secp256 curve.

> fn verify(self: KeyPair<u256, Secp256Point>, message_hash: H, signature: (u256, u256)) -> bool

## Example

```rust
use snforge_std::signature::key_pair::{KeyPair, KeyPairTrait, SignerTrait, VerifierTrait};
use snforge_std::signature::secp256_curve::{Secp256CurveKeyPairImpl, Secp256CurveSignerImpl, Secp256CurveVerifierImpl};
use starknet::secp256k1::{Secp256k1Impl, Secp256k1Point, Secp256k1PointImpl};

> use core::starknet::SyscallResultTrait;

Reviewer: Imo this is to wordy and **very** hard to maintain, just give an example how to use it and that is enough.
